# CMash-reproducibles
For reproducibility of [Cmash](https://github.com/dkoslicki/CMash) tasks/issues:
1. [K-mer truncation](https://github.com/dkoslicki/CMash/issues/20)

## Install dependencies  
1. Install conda or [miniconda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/) (version >= 4.6)  
2. Clone this repo
```
git clone https://github.com/KoslickiLab/CMASH-reproducibles.git
```
3. Go to "src" folder, run cmd `bash install_dependency.sh`. This will install all necessary tools and download needed reources in the "src" folder, it may take several minuates. To specify location for Conda environment instead of the default location, please add `-p <abs_path_to_folder>` with this cmd, or add `-p default` to install them in the original conda folder.  

## Reproducibility
1. To reproduce reuslts of task1 (kmer truncation), please:  
   - go to folder "task1_k-mer_truncation/example"
   - run the script in "src" folder (you may use -h to see the help information):  
   ```
   bash <path_2_src_folder>/rep1_main_kmer_truncation.sh -q meta_high_CI.txt  -r ref_high_CI.txt -k 60 -c 5-60-5 -d <absolute_path_2_conda_folder> -g <absolute_path_2_this_repo>
   ```

## Task1: how would truncation of long k-mers affect containment index?
Comparing the containment index generated by truncated k-mers (prefix of original long k-mers) and original k-mers of a large k (such as 60).

#### Input file:
1. ref file: a file containing absolute path of all reference data in analysis
2. query file: a file containing absolute path of all query metagenome data in analysis (must be unzipped, as CAMISIM can't work with zipped file)

#### Output file:
1. CAMISIM and BBMap simulated reads in fastq format
2. Tenery search tree of a range of k-mer database
3. Estimated / ground_truch / truncated containment index (stored in csv file) of the input metagenome files
4. Summary plot of the comparison

#### Example
[Click here](https://github.com/KoslickiLab/CMASH-reproducibles/tree/master/task1_K-mer_truncation/example) for an example input and output.  
Input:
1. ref file is "path_ref.txt"
2. query file is "path_meta.txt"

Output:
1. estimated_CI: store the estimated containment index between metagenome and reference datasets
2. ground_truth_CI: store the ground truch containment index between metagenome and reference datasets
3. truncated_CI: store the containment index estimated by truncated k-mer
4. ref_df: store the tenery search tree of each k size (intentially left empty here as the file size is large)
5. summary: the summary plot



